File: Semester 4/Advanced Programming/RpcProtocolJava/ApplicationNetworking/src/main/java/app/network/ClientWorker.java
package app.network;

import app.model.Cursa;
import app.model.Rezervare;
import app.model.User;
import app.network.dto.DestinationDateDto;
import app.services.IObserver;
import app.services.IServices;
import app.services.ServiceException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.time.LocalDateTime;
import java.util.List;

public class ClientWorker implements Runnable, IObserver {
    private IServices service;
    private Socket socket;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientWorker(IServices service, Socket client){
        this.service = service;
        this.socket=client;
        try{
            output =new ObjectOutputStream(socket.getOutputStream());
            output.flush();
            input =  new ObjectInputStream(socket.getInputStream());
            connected = true;
        }catch (IOException ex){
            ex.printStackTrace();
        }
    }
    @Override
    public void run() {
        while(connected){
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse(response);
                }
            }catch (IOException ex){
                ex.printStackTrace();
            }catch (ClassNotFoundException ex){
                ex.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }//    pentru?
        }
        try{
            input.close();
            output.close();
            socket.close();
        }catch (IOException ex){
            System.out.println("Eroare in clientWorker "+ex);
        }
    }

    private synchronized void sendResponse(Response response) throws IOException {
        output.writeObject(response);
        output.flush();
    }
    private Response createErrorResponse(Exception ex){
        return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();

    }
    private static Response okResponse = new Response.Builder().type(ResponseType.OK).build();
    private Response handleRequest(Request request) {
        Response response = null;
        switch (request.type()){
            case LOGIN:
                System.out.println("Login request received...");
                User user = (User)request.data();
                try{
                    service.login(user,this);
                    return okResponse;
                }catch (ServiceException ex){
                    connected = false;
                    return createErrorResponse(ex);
                }

            case LOGOUT:
                System.out.println("Logout request received...");
                try {
                    service.logout(this);
                } catch (ServiceException e) {
                    e.printStackTrace();
                }
                connected=false;
                break;

            case GETALLCURSE:
                System.out.println("GETALLCURSE request received ...");
                try{
                    List<Cursa> curse = service.getAllCurse();
                    return new Response.Builder().type(ResponseType.OK).data(curse).build();
                }catch (ServiceException ex){
                    return createErrorResponse(ex);
                }

            case FindByDestinationAndDate:
                System.out.println("FindByDestinationAndDate request received ....");
                try{
                    DestinationDateDto dto = (DestinationDateDto) request.data();
                    String destination  = dto.getDestination();
                    LocalDateTime date = dto.getDate();
                    Cursa cursa = service.findByDestinationAndDate(destination,date);
                    return new Response.Builder().type(ResponseType.OK).data(cursa).build();
                } catch (ServiceException e) {
                    return createErrorResponse(e);
                }
            case GetAllBookings:
                System.out.println("GetAllBookings request received ...");
                Integer cursaId = (Integer)request.data();
                try{
                    List<Rezervare> rezervari = service.getAllBookings(cursaId);
                    return new Response.Builder().type(ResponseType.OK).data(rezervari).build();
                } catch (ServiceException e) {
                    return createErrorResponse(e);
                }
            case MakeBooking:
                System.out.println("MakeBooking request received ...");
                List<Rezervare> rezervari = (List<Rezervare>)request.data();
                try{
                    service.makeBooking(rezervari);
                    return okResponse;
                }catch (ServiceException ex){
                    return createErrorResponse(ex);
                }
        }
        return response;
    }

    @Override
    public synchronized void seatsDecremented(Cursa cursa) throws ServiceException {
        Response response = new Response.Builder().type(ResponseType.SeatsDecremented).data(cursa).build();
        try{
            sendResponse(response);
        } catch (IOException e) {
            throw new ServiceException("SeatsDecremented error "+e);
        }

    }

    @Override
    public synchronized void newBookings(List<Rezervare> rezervari) throws ServiceException {
        Response response = new Response.Builder().type(ResponseType.NewBookings).data(rezervari).build();
        try{
            sendResponse(response);
        } catch (IOException e) {
            throw new ServiceException("NewBookings error "+e);
        }

    }
}

File: Semester 4/Advanced Programming/RpcProtocolC#/Networking/ClientWorker.cs
using Domain;
using services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
namespace network
{
    public class ClientWorker : IObserver
    {
        private IServices service;
        private TcpClient client;
        private volatile Boolean connected;
        private NetworkStream stream;
        private IFormatter formatter;
        public ClientWorker(IServices service, TcpClient client)
        {
            this.service = service;
            this.client = client;
            try
            {
                stream = client.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }catch(Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

        }
        public virtual void newBookings(List<Rezervare> rezervari)
        {
            Response response = Response.Builder.create().Type(ResponseType.NewBookings).Data(rezervari);
            sendResponse(response);
        }

        public virtual  void seatsDecremented(Cursa cursa)
        {
            Response response = Response.Builder.create().Type(ResponseType.SeatsDecremented).Data(cursa);
            sendResponse(response);
        }
        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    Console.WriteLine("ClientWorker: request received "+((Request)request).type);

                    Response response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                client.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("CLientWorker: sending Response : " + response.type);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }
            Console.WriteLine("Response was sent.");
        }

        private Response handleRequest(Request request)
        {
            Response response = null;
            switch (request.type)
            {
                case RequestType.LOGIN:
                    Console.WriteLine("login request received");
                    User user = (User)request.data;
                    try
                    {
                        lock (service)
                        {
                            service.login(user, this);
                        }
                        Console.WriteLine("User logged in, about to create response");
                        return Response.Builder.create().Type(ResponseType.OK);
                    }
                    catch (ServiceException e)
                    {
                        connected = false;
                        return Response.Builder.create().Type(ResponseType.Error).Data(e.Message);
                    }

                case RequestType.GetAllCurse:
                    Console.WriteLine("GetAllCurse request received");
                    try
                    {
                        IList<Cursa> curse;
                        lock (service)
                        {
                            curse = service.getAllCurse();
                        }
                        return Response.Builder.create().Type(ResponseType.OK).Data(curse);
                    }catch(ServiceException e)
                    {
                        return Response.Builder.create().Type(ResponseType.Error).Data(e.Message);
                    }
                case RequestType.FindCursa:
                    Console.WriteLine("FindCursa request received");
                    try
                    {
                        Cursa cursa;
                        lock (service)
                        {
                            DestinationDateDto dto = (DestinationDateDto)request.data;
                            string destination = dto.getDestination();
                            DateTime date = dto.getDate();
                            cursa = service.findByDestinationAndDate(destination, date);
                            Console.WriteLine("Retreived cursa from service");
                        }
                        return Response.Builder.create().Type(ResponseType.OK).Data(cursa);
                    }catch(ServiceException e)
                    {
                        return Response.Builder.create().Type(ResponseType.Error).Data(e.Message);
                    }
                case RequestType.LOGOUT:
                    Console.WriteLine("Logout request received");
                        lock (service)
                        {
                            service.logout(this);
                        }
                    connected = false;
                    return null;
                case RequestType.GetAllBookings:
                    Console.WriteLine("GetAllBookings request received");
                    try
                    {
                        List<Rezervare> bookings = null;
                        lock (service)
                        {
                            bookings = service.getAllBookings((int)request.data);
                        }
                        return Response.Builder.create().Type(ResponseType.OK).Data(bookings);
                    }catch  (ServiceException ex)
                    {
                        return Response.Builder.create().Type(ResponseType.Error).Data(ex.Message);
                    }
                case RequestType.MakeBooking:
                    Console.WriteLine("MakeBooking request received");
                    try
                    {
                        lock (service)
                        {
                            service.makeBooking((List<Rezervare>)request.data);
                        }
                        return Response.Builder.create().Type(ResponseType.OK);
                    }
                    catch(ServiceException ex)
                    {
                        return Response.Builder.create().Type(ResponseType.Error).Data(ex.Message);
                    }
            }
            return response;
        }
    }
}
